### ISSUE TITLE
Allow use of enviroment variables in profiles when using profiles_yml_filepath

### ISSUE THREAD

**Issue by DanMawdsleyBA** [astronomer/astronomer-cosmos#468]
I'm looking to build a dag which uses environment variables in the profiles when using a yml path. Something like the below:

```
dbt_envs= {
    'user': 'test',
    'private_key': <lookup private key>
}

profile_config = ProfileConfig(
    profile_name='dbt_project',
    target_name='snowflake_dev',
    profiles_yml_filepath='/usr/local/airflow/dags/dbt/profiles.yml',
    env=dbt_envs
)
```

With the profile yml referencing it like:
```
    snowflake_dev:
      type: snowflake
      account: test_account.eu-west-1
      user: "{{ env_var('USER') }}"
      private_key: "{{ env_var('PRIVATE_KEY') }}"
```

**Comment by tatiana** [comment]
Hi @DanMawdsleyBA, Thanks for your message!

Some questions:
1) Which `execution_mode` are you using?
2) Are you setting the environment variables `USER` and `PRIVATE_KEY` when running Airflow, or would the expectation be that Cosmos should set them at a process level?

If the environment variables are available for the Airflow process, a way to forward them to the underlining Cosmos/Airflow operators is to use:
```
`operator_args={"append_env": True}`
```

**Comment by DanMawdsleyBA** [comment]
Thanks for your response so I would be looking to build a dbt dag that would run all models and tests so running `dbt build`
So in the python dag I have lookup the aws secret using `service_client.get_secret_value(SecretId='secret_value')` this retrieves all the connection details. 

So if set the operator_args as above could I do something like this:
in the dag have:
```
user='user'
private_key=service_client.get_secret_value(SecretId='secret_value')

my_cosmos_dag = DbtDag(
    project_config=ProjectConfig(dbt_project_path='/usr/local/airflow/dags/oprn_crw_ops_planning/dbt'),
    profile_config=profile_config,
    operator_args={"append_env": True},
    # normal dag parameters
    schedule_interval='@daily',
    start_date=datetime(2023, 1, 1),
    catchup=False,
    dag_id=f'{Path(__file__).stem}_{env}',
)
```
in the profile have the below and it would be populated with the value user and private key from the dag:
```
    snowflake_dev:
      type: snowflake
      account: test_account.eu-west-1
      user: "{{ env_var('USER') }}"
      private_key: "{{ env_var('PRIVATE_KEY') }}"
```
Is that correct?

**Comment by tatiana** [comment]
Hi @DanMawdsleyBA , sorry for the delay getting back to you.

Do you know if this latest change worked?

You may have to capitalize `user` for this to work:
```
dbt_envs{
    'USER': 'user',
    'PRIVATE_KEY': service_client.get_secret_value(SecretId='secret_value')
}
```

If there are errors, could you share the logs?

The problem you may be having if Cosmos is using `load_method=LoadMode.DBT_LS` is that the environment variable is only being set for the execution of the operators, but not at DAG processing time.

If that is the case, something that may solve the issue is to do this in your DAG:

```
import os
os.environ["USER"] = 'user'
os.environ["PRIVATE_KEY"]  = service_client.get_secret_value(SecretId='secret_value')
```

**Comment by DanMawdsleyBA** [comment]
Hi @tatiana the airflow logs themselves aren't useful as they just display no tasks but the dag parses. When I run a similar version locally using the debugger I get this so I can see the environment variables are loaded for user and private key. But it hits an error later on which looks to do with the sources (This is a working dbt project both on my local and airflow using a bash command) See below error (I'm using dbt-snowflake v1.52 and astronomer-cosmos v1.0.5):

![image](https://github.com/astronomer/astronomer-cosmos/assets/134394384/c09f6637-0a59-423c-947a-38290d65ef62)
Looking at the graph file here does there need to allow for sources but not create them as tasks?

![image](https://github.com/astronomer/astronomer-cosmos/assets/134394384/8c95c8d8-cc4f-4838-868a-e06933989e09)

**Comment by tatiana** [comment]
@DanMawdsleyBA I agree the current behavior is misleading.

As you noticed in the source code, at the moment, Cosmos does not create tasks for DBT sources. Even though the log message says "error", it is more like a warning since the code skips those tasks, continuing to build the DAG with other supported resource types.

I just checked the [dbt Sources](https://docs.getdbt.com/docs/build/sources) documentation. How could we improve Cosmos?

Would you like to see each source materialized as an Airflow task?

What would be the expected behavior of those tasks?

Perhaps, if users set `TestBehavior.AFTER_ALL`, we could have a similar behavior to what we do with models, having a Task Group that joins the source task and that source test.

Looping in @jlaneve who can give more ideas and suggestions.

**Comment by DanMawdsleyBA** [comment]
@tatiana I guess you include sources as tasks even if it's as simple task as checking if the table exists. A future feature perhaps could add a setting like freshness to see if that table has been loaded within a set sla period.

When I upload the dag should I be able to see all of the tasks (Even if the connection isn't setup)? Currently I'm unable to get any tasks to be created and not getting any logs or errors from Airflow. Not sure if you have any tips on debugging?

**Comment by tatiana** [comment]
I like your proposal, @DanMawdsleyBA , could you create a dedicated Github issue with this suggestion for creating a task per source, with the suggested behaviour?

Based on the code you shared before, it seems you're using a [custom profile](https://astronomer.github.io/astronomer-cosmos/profiles/index.html#using-your-own-profiles-yml-file), using `profiles_yml_filepath='/usr/local/airflow/dags/dbt/profiles.yml`. In this scenario, Cosmos does not rely on Airflow connections. Is this file available in your remote deployment?

Where/how are you running your remote Airflow service?  The conversion between a dbt project into an Airflow DAG happens during Airflow DAG parsing. Airflow names these logs as DAG processor logs ([more on Airflow logs config](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#logging)). The logs' location will likely depend on your remote setup.

Once you find where the logs are being stored remotely, setting the logging level to DEBUG in your remote Airflow instance can be helpful. If you're using environment variables, a way of accomplishing this is:
```
export AIRFLOW__CORE__LOGGING_LEVEL=DEBUG
```

**Comment by jlaneve** [comment]
> @DanMawdsleyBA I agree the current behavior is misleading.
> 
> 
> 
> As you noticed in the source code, at the moment, Cosmos does not create tasks for DBT sources. Even though the log message says "error", it is more like a warning since the code skips those tasks, continuing to build the DAG with other supported resource types.
> 
> 
> 
> I just checked the [dbt Sources](https://docs.getdbt.com/docs/build/sources) documentation. How could we improve Cosmos?
> 
> 
> 
> Would you like to see each source materialized as an Airflow task?
> 
> 
> 
> What would be the expected behavior of those tasks?
> 
> 
> 
> Perhaps, if users set `TestBehavior.AFTER_ALL`, we could have a similar behavior to what we do with models, having a Task Group that joins the source task and that source test.
> 
> 
> 
> Looping in @jlaneve who can give more ideas and suggestions.
> 
> 
> 
> 

I've spent a lot of time thinking about this and talking to users, and I think the right approach is outlined in this issue: https://github.com/astronomer/astronomer-cosmos/issues/427. Everyone's going to want different behavior with their sources (some people don't want them, some people want them as sensors, custom operators, etc) so a generating function seems to be the most flexible.

**Comment by DanMawdsleyBA** [comment]
> I like your proposal, @DanMawdsleyBA , could you create a dedicated Github issue with this suggestion for creating a task per source, with the suggested behaviour?
> 
> Based on the code you shared before, it seems you're using a [custom profile](https://astronomer.github.io/astronomer-cosmos/profiles/index.html#using-your-own-profiles-yml-file), using `profiles_yml_filepath='/usr/local/airflow/dags/dbt/profiles.yml`. In this scenario, Cosmos does not rely on Airflow connections. Is this file available in your remote deployment?
> 
> Where/how are you running your remote Airflow service? The conversion between a dbt project into an Airflow DAG happens during Airflow DAG parsing. Airflow names these logs as DAG processor logs ([more on Airflow logs config](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#logging)). The logs' location will likely depend on your remote setup.
> 
> Once you find where the logs are being stored remotely, setting the logging level to DEBUG in your remote Airflow instance can be helpful. If you're using environment variables, a way of accomplishing this is:
> 
> ```
> export AIRFLOW__CORE__LOGGING_LEVEL=DEBUG
> ```

Thanks for your help! I was able to find the logs for when dbt parses the dag and saw that no nodes were found even though both the profile, project are all in place. When I ran it locally I saw there was 33 nodes. I think there seems to be some sort of path issue where the `dbt ls` is possibly not running in the correct path. In my case on airflow I set project path to: '/usr/local/airflow/dags/oprn_crw_ops_planning/dbt' but the logs show the command here 
``/usr/local/airflow/.local/bin/dbt ls --output json --profiles-dir /usr/local/airflow/dags/oprn_crw_ops_planning/dbt --profile dbt_crew --target snowflake`` . So it might be running in the wrong path. I saw you made some changes previously here: https://github.com/astronomer/astronomer-cosmos/issues/411 it might be possible that fixes this issue as you specify the project path

**Comment by tatiana** [comment]
Hey Dan, the change you mentioned in the other ticket was released as part of 1.0.5 and 1.1.0a1 - you could try it out.

I just saw your message in the other ticket. was the issue running `dbt deps` before `dbt ls`?

**Comment by DanMawdsleyBA** [comment]
> Hey Dan, the change you mentioned in the other ticket was released as part of 1.0.5 and 1.1.0a1 - you could try it out.
> 
> I just saw your message in the other ticket. was the issue running `dbt deps` before `dbt ls`?

Yeah it was the issue of not running `dbt deps`. When I removed all of the packages I was able to get the dag to parse correctly

**Comment by tatiana** [comment]
@DanMawdsleyBA I'm closing this ticket since we have a ticket to add `dbt ls` in the backlog: #445. I'm very sorry it took so long to troubleshoot this problem. We'll improve this. If you are available and could help with the project, we'd also appreciate any contributions.
